pages/api/autocomplete.ts

```diff
@@ -25,36 +25,92 @@ graph.updateEachNodeAttributes((node, attr) => {
 });
 
 type QueueItem = {
-  node: string;
-  characters: string;
-  toNext: boolean;
+  priority: number;
+  data: {
+    node: string;
+    characters: string;
+    toNext: boolean;
+  };
 };
 
-function bfsRecursionToNextData(queue: QueueItem[]) {
-  const queueItem = queue.shift();
-  //console.log(graph.getNodeAttribute(queueItem?.node, 'c'));
-  if (graph.getNodeAttribute(queueItem?.node, 'visited')) {
+class PriorityQueue {
+  items: QueueItem[];
+  constructor() {
+    this.items = [];
+  }
+  enqueue(queueItem: QueueItem) {
+    let contain = false;
+    for (let i = 0; i < this.items.length; i++) {
+      if (this.items[i].priority > queueItem.priority) {
+        this.items.splice(i, 0, queueItem);
+        contain = true;
+        break;
+      }
+    }
+    if (!contain) {
+      this.items.push(queueItem);
+    }
+  }
+  dequeue() {
+    if (this.isEmpty()) {
+      return;
+    }
+    return this.items.shift();
+  }
+  front() {
+    if (this.isEmpty()) {
+      return;
+    }
+    return this.items[0];
+  }
+  isEmpty() {
+    return this.items.length === 0;
+  }
+}
+
+function bfsRecursionToNextData(queue: PriorityQueue) {
+  const queueItem = queue.dequeue();
+  //console.log(graph.getNodeAttribute(queueItem?.data?.node, 'c'));
+  if (graph.getNodeAttribute(queueItem?.data?.node, 'visited')) {
     return;
   }
-  graph.setNodeAttribute(queueItem?.node, 'visited', true);
-  const data = graph.getNodeAttribute(queueItem?.node, 'd');
+  graph.setNodeAttribute(queueItem?.data?.node, 'visited', true);
+  const data = graph.getNodeAttribute(queueItem?.data?.node, 'd');
   if (typeof data !== 'undefined') {
+    graph.forEachOutNeighbor(queueItem?.data?.node, (neighbor) => {
+      queue.enqueue({
+        priority:
+          (queueItem?.priority ?? 0) +
+          100 +
+          graph.getNodeAttribute(neighbor, 'c').length,
+        data: {
+          node: neighbor,
+          characters: '',
+          toNext: true,
+        },
+      });
+    });
     return data;
   } else {
-    graph.forEachOutNeighbor(queueItem?.node, (neighbor) => {
-      queue.push({
-        node: neighbor,
-        characters: '',
-        toNext: true,
+    graph.forEachOutNeighbor(queueItem?.data?.node, (neighbor) => {
+      queue.enqueue({
+        priority:
+          (queueItem?.priority ?? 0) +
+          graph.getNodeAttribute(neighbor, 'c').length,
+        data: {
+          node: neighbor,
+          characters: '',
+          toNext: true,
+        },
       });
     });
   }
   return;
 }
 
-function bfsRecursion(queue: QueueItem[]) {
-  const queueItem = queue.shift();
-  if (typeof queueItem === 'undefined') {
+function bfsRecursion(queue: PriorityQueue) {
+  const queueItem = queue.dequeue();
```

Comment:
In order to remove all of the `queueItem`s, could a type guard be implemented at the top level of this function and provide another variable for use within the function?

Reply:
My mistake, I missed a ? character! Instead of the optional `queueItems` could this be effectively unwrapped at a high level for use below?

Reply:
Like instead of passing in `queue`, passing in `queue.dequeue()` and `queue.enqueue`?

Reply:
I'm not sure the best way to do this in TypeScript, the way I might write something like this in Swift is
```if let queueItem = queue.dequeue() {
          queueItem is always non-nil here
          } else {
          handle the nil case
          }
 ```
 
 The idea would be to guarantee that queueItem is never nil, so we can use `queueItem` instead of the optional `queueItem?`. Not a huge deal, and I think it's fine as it, but could be something to consider if this type of thing is doable in TypeScript.

Reply:
Good idea, that let me remove all the `?.`s. Done